gateway/handler.go

```diff
@@ -727,6 +733,45 @@ func handleServiceWorkerRegistration(r *http.Request) (err *ErrorResponse) {
 	return nil
 }
 
+// handleIpnsB58mhToCidRedirection redirects from /ipns/b58mh to /ipns/cid in
+// the most cost-effective way.
+func handleIpnsB58mhToCidRedirection(w http.ResponseWriter, r *http.Request) bool {
+	if w.Header().Get("Location") != "" {
+		// Ignore this if there is already a redirection in place. This happens
+		// if there is a subdomain redirection. In that case, the path is already
+		// converted to CIDv1.
+		return false
+	}
+
+	pathParts := strings.Split(r.URL.Path, "/")
+	if len(pathParts) < 3 {
+		return false
+	}
+
+	if pathParts[1] != "ipns" {
+		return false
+	}
+
+	s := pathParts[2]
+
+	// Similarly to peer.Decode, check the prefix first as it is  less computationally
+	// expensive. Base58 encoded peer IDs will start with a '1'.
+	if !strings.HasPrefix(s, "1") {
+		return false
+	}
+
+	// Decode the base58 encoded sha256 or identity multihash.
+	m, err := multihash.FromB58String(s)
+	if err != nil {
+		return false
+	}
```

Comment:
@lidel `peer.Decode` will also work with b58mh, leading to infinite redirects. That's what I was trying to avoid here. How do you recommend going forward? Decoding it __always__ and comparing it with the string we already have?